### ISSUE TITLE
Many internal anchors for subsections are not being generated.

### ISSUE THREAD

**Issue by thundergnat** [Raku/doc-website#153]
There are many cross references throughout the docs that link to internal subsections. The subsections do not have anchors generated, so fail with 404.

There are many. A few examples:

Page: [Syntax](https://docs.raku.org/language/syntax.html), at bottom of page, link to [extended identifier](https://docs.raku.org/syntax/identifiers.html#Extended_identifiers)  fails.

Page: [Cool](https://docs.raku.org/type/Cool.html), method EVAL section links to [subroutine form](https://docs.raku.org/language/independent-routines.html#routine_EVAL). Fails with 404.

In [Perl to Raku guide - operators](https://docs.raku.org/language/5to6-perlop.html), Quote and quote-like operators subsection, [zen-slice](https://docs.raku.org/language/subscripts.html#index-entry-Zen_slices) doesn't 404, but it doesn't link to the zen-slice subsection either. 

May be better to just automatically generate anchors for each subsection. Would make it less problematic to track which subsections need an anchor if they all do by default. 

(And would make it easier to link to a specific subsection of the documentation in general.)

**Comment by finanalyst** [comment]
There is a link checker that picks up many (most/all) of these 404 errors. It was decided not to put a link to these errors in the Production site. You can see the results of the test at [new-raku.finanalyst.org](https://new-raku.finanalyst.org/error-report.html). 

There are several reasons the errors have occurred, eg.,
- Changes in the source of the documentation, the old site was not being automatically generated for several months. It contained quite a few 404s too.
- Changes made to eliminate other errors (eg. mismatch of visible header) For example, links to `callsame` and `callwith` in Operators 404 because when the documentation source was written, the links were made to empty `X<>` entries (index entries with only an anchor, but no text for the glossary) and not to the heading. This resulted in the browser not showing the heading. The problem was fixed and the `X<>` code seems to have been deleted. But the links were not changed. 
- Systematic changes due to new tooling. Such mistakes are entirely my bad, but help is needed to identify whether a 404 is due to a change in the source documentation,

**Comment by cfa** [comment]
Related to this,

> May be better to just automatically generate anchors for each subsection. Would make it less problematic to track which subsections need an anchor if they all do by default.

the links generated for anchored headings are currently incorrect across the board.

For example, take a look at [Any](https://docs.raku.org/type/Any.html) (though any page will do). Clicking on the *Methods* heading will take you back to the top of the page because of the incorrect link target,

```html
<h1 id="Methods" class="raku-h1"><a href="#class_Any" class="u" title="go to top of document">Methods</a></h1>
```

This should read `<a href="#Methods"...>`.

The issue may be more obvious if you scroll to the bottom of the page and try the `sum` link,

```html
<h2 id="method_sum" class="raku-h2"><a href="#class_Any" class="u" title="go to top of document">method sum</a></h2>
```

Notice the same `#class_Any` target (rather than `#method_sum`).

**Comment by finanalyst** [comment]
@cfa The analysis is not right - I think. Look at #93 
@lizmat pointed out that the `<a>` on titles was included in the original HTML 1.0 rendering in order to get a TOC. So `<a href="#class_Any` is pointing back to the top of the page, where the TOC was located as a paragraph of links. 
With the re-tooling, we have more control over how the TOC is generated (its in a closable sidebar) and how the titles are rendered.
I like @2colours idea where scrolling down the sidebar entries brings a header into the visibile browser window. 

The issue with the 404 subsections is more complicated and is related to the way title text is mangled to get HTML ids. There was a change in the id generator in Documentable which made some links unreachable.

**Comment by 2colours** [comment]
> I like @2colours idea where scrolling down the sidebar entries brings a header into the visibile browser window.

Sorry but I think you got this backwards: actually, hovering the header entry highlights the corresponding ToC link, so that you know where you are in the broader context of things. However, if you want it to be the other way around, that's also possible (and less challenging actually - less data to keep track of, and no sidebar challenge).

**Comment by cfa** [comment]
> @cfa The analysis is not right - I think. Look at #93 @lizmat pointed out that the `<a>` on titles was included in the original HTML 1.0 rendering in order to get a TOC. So `<a href="#class_Any` is pointing back to the top of the page, where the TOC was located as a paragraph of links. With the re-tooling, we have more control over how the TOC is generated (its in a closable sidebar) and how the titles are rendered. I like @2colours idea where scrolling down the sidebar entries brings a header into the visibile browser window.

I'm not sure I follow. I'm saying that these links shouldn't point to the top of the page at all (context is lost and behaviour is non-standard).

The current implementation unexpectedly navigates the user away from the section that they were looking at. It sounds like this is a byproduct of changes to the TOC but even so, the link targets as they stand yield surprising (and unhelpful) results.

**Comment by cfa** [comment]
P.S. a common (and standard) use case is to click a heading to explicitly navigate to its anchor, then copying the URL.

For example, I may want to click the `sum` heading on `Any` to visit `#method_sum`, then copy the URL and share so that you can quickly navigate to the correct part of the documentation.

(This currently requires re-navigating to `sum` in the sidebar.)

**Comment by finanalyst** [comment]
@cfa I understand your use case, and I am working to implement that too. In the old site, each heading had two `<a>` tags, one pointing to itself, and one to the top of the page.

The one pointing to the header itself would appear only upon mouse over and a 'hook' character would become visible to the right of the header, and clicking on it would provide the URL to copy.

I did not know about the use case for copying the URL. In addition, headers were often marked up with `X<>` which also create `<a>` tabs. So a header often had three `<a>` tabs. And they were nested. This led to some headers having different behaviours. 

To simplify this, I changed the rendering of headers, and eliminated the `<a>` that produced the URL.

In another issue, we have discussed removing the link back to the top of the web page (eg. the `#class_Any` you referenced). @2colours has started this work.
